help/learn/Custom-Machine-Learning.md

```diff
@@ -0,0 +1,132 @@
+# Custom ML models
+ 
+ Datagrok supports three ML engines out of the box: H2O, Caret, Chemprop. In addition, the platform allows users to build their own customizable ML models. This capability provides full access and control to algorithms available in ML libraries in any of the supported languages (such as Python, R, and Julia). The users may construct and configure any chosen model in their custom data pipeline.   
+ 
+ Custom model functionality is implemented as a two-step process using train and apply functions. Train function is used to build/train a model based on provided features and a target variable. The trained model is then stored as an object inside a given directory. Apply function receives previously saved model and applies it to provided feature columns. The resulting predictions are returned in a form of a single column in a dataframe.  
+ 
+Custom models utilize Predictive Model interface alongside other out-of-the-box solutions. Once implemented, custom models can be chosen from the list of model engines.
+
+
+## Train
+ 
+Header Parameters
+
+- #meta.mlname: CustomName – Name of the custom train function.
+- #meta.mlrole: train –  Action (train or apply).
+- #description: Custom ML train function for KNN algorithm – Description of the function.
+- #language: r – Language (Python, R, Julia).
+ 
+Input Data Parameters
+
+- #input: dataframe df – dataframe for training.
+- #input: string predict_column – List of features/column names separated by a comma.
+ 
+Input Model Parameters
+
+- #input: int parm1 {category: group1} - Parameter 1 to build a model. Category is used for input parameters grouping within the UI.
+ 
+Output Model Parameters
+
+- #output: blob modelName – Trained model object name. Complete trained model with an absolute path address is stored as a blob object to be retrieved and applied by the Apply function.
+ 
+## Apply
+ 
+Header Parameters
+
+- #meta.mlname: CustomName – Name of the custom apply function (should match corresponding train function name).
+- #meta.mlrole: apply –  Action (train or apply).
+- #description: Custom ML apply function for KNN algorithm – description of the function.
+- #language: Python – Language (Python, R, Julia).
+ 
+Input Model Parameters
+
+- #input: blob model – Trained model object name. Complete trained model with an absolute path address saved by the Train function.
+
+
+ 
+Input Data Parameters
+
+- #input: dataframe df - dataframe for prediction (contains only prediction features).
+- #input: string namesKeys – optional list of original features/column names separated by comma.
+- #input: string namesValues – optional list of new features/column names separated by comma. If both namesKeys and namesValues are supplied, namesKeys will replace corresponding namesValues feature names before accessing the dataframe.
+ 
+Output Parameters
+
+- #output: dataframe data_out – single-column dataframe of predicted values.
+ 
+## Example
+
+#### Train Function
```

Comment:
Let's change the heading level step by step, from the second level (##) to the third (###)